src/aws_encryption_sdk/materials_managers/__init__.py

```diff
@@ -40,38 +48,254 @@ class EncryptionMaterialsRequest(object):
+
+    def add_data_encryption_key(self, data_encryption_key, keyring_trace):
+        # type: (Union[DataKey, RawDataKey], KeyringTrace) -> None
+        """Add a plaintext data encryption key.
+
+        .. versionadded:: 1.5.0
+
+        :param RawDataKey data_encryption_key: Data encryption key
+        :param KeyringTrace keyring_trace: Trace of actions that a keyring performed
+          while getting this data encryption key
+        :raises AttributeError: if data encryption key is already set
+        :raises InvalidKeyringTraceError: if keyring trace does not match generate action
+        :raises InvalidKeyringTraceError: if keyring trace does not match data key provider
+        :raises InvalidDataKeyError: if data key length does not match algorithm suite
+        """
+        self._add_data_encryption_key(
+            data_encryption_key=data_encryption_key,
+            keyring_trace=keyring_trace,
+            required_flags={KeyringTraceFlag.WRAPPING_KEY_GENERATED_DATA_KEY},
+        )
+
+    def add_encrypted_data_key(self, encrypted_data_key, keyring_trace):
+        # type: (EncryptedDataKey, KeyringTrace) -> None
+        """Add an encrypted data key with corresponding keyring trace.
+
+        .. versionadded:: 1.5.0
+
+        :param EncryptedDataKey encrypted_data_key: Encrypted data key to add
+        :param KeyringTrace keyring_trace: Trace of actions that a keyring performed
+          while getting this encrypted data key
+        :raises AttributeError: if data encryption key is not set
+        :raises InvalidKeyringTraceError: if keyring trace does not match generate action
+        :raises InvalidKeyringTraceError: if keyring trace does not match data key encryptor
+        """
+        if self.data_encryption_key is None:
+            raise AttributeError("Data encryption key is not set.")
+
+        if KeyringTraceFlag.WRAPPING_KEY_ENCRYPTED_DATA_KEY not in keyring_trace.flags:
+            raise InvalidKeyringTraceError("Keyring flags do not match action.")
+
+        if keyring_trace.wrapping_key != encrypted_data_key.key_provider:
+            raise InvalidKeyringTraceError("Keyring trace does not match data key encryptor.")
+
+        self._encrypted_data_keys.add(encrypted_data_key)
+        self._keyring_trace.append(keyring_trace)
+
+    def add_signing_key(self, signing_key):
+        # type: (bytes) -> None
+        """Add a signing key.
+
+        .. versionadded:: 1.5.0
+
+        :param bytes signing_key: Signing key
+        :raises AttributeError: if signing key is already set
+        :raises SignatureKeyError: if algorithm suite does not support signing keys
+        """
+        if self.signing_key is not None:
+            raise AttributeError("Signing key is already set.")
+
+        if self.algorithm.signing_algorithm_info is None:
+            raise SignatureKeyError("Algorithm suite does not support signing keys.")
+
+        self._setattr("signing_key", signing_key)
```

Comment:
We should check that the signatureCurve for the key is the same as the suite....

https://github.com/awslabs/aws-encryption-sdk-javascript/issues/130

Reply:
I agree, but digging into this, I think the right way to address this will be my adding this verification into the `Signer` and `Verifier` classes that we use to abstract all of the signing and verification away. For the time being, I'm just going to create an instance of each to "verify" the key and address the actual verification in https://github.com/aws/aws-encryption-sdk-python/issues/167.